File: eos5/swath.c
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 * Copyright by The HDF Group.                                               *
 * Copyright by the Board of Trustees of the University of Illinois.         *
 * All rights reserved.                                                      *
 *                                                                           *
 * This file is part of HDF5.  The full HDF5 copyright notice, including     *
 * terms governing use, modification, and redistribution, is contained in    *
 * the files COPYING and Copyright.html.  COPYING can be found at the root   *
 * of the source code distribution tree; Copyright.html can be found at      *
 * http://hdfgroup.org/HDF5/doc/Copyright.html.  If you do not have          *
 * access to either file, you may request a copy from help@hdfgroup.org.     *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

/*
 * created by:  Isayah Reed <ireed@hdfgroup.org>
 *		May  4, 2011
 * 
 *  This is a simple program that demonstrates how to use HDF-EOS5 to add CF 
 *  attributes to an h5 file. Creates a swath called Swath1 and adds the 
 *  temperature dataset with 2 dimensions: latitude and longitude. The
 *  temperature dataset has the CF attributes: units, long_name, _FillValue,
 *  coordinates, valid_min, valid_max, valid_range, scale_factor, add_offset. 
 *  Latitude and longitude have the CF attributes units and long_name. Outputs 
 *  data to general.h5
 *
 * Compile:
 *  $(HDF5_DIR)/bin/h5cc swath.c -I$(EOS_DIR)/include -L$(EOS_DIR)/lib -lhe5_hdfeos -lGctp
 *
*/


#include "HE5_HdfEosDef.h"

#define NX1     180
#define NY1     360

#define TEMP "temp"
#define LAT "lat"
#define LON "lon"

#define UNITS "Units"
#define FILLVALUE "_FillValue"
#define LONGNAME "long_name"
#define COORDINATES "coordinates"
#define STDNAME "standard_name"
#define ADDOFFSET "add_offset"
#define SCALEFACTOR "scale_factor"
#define VALIDMIN "valid_min"
#define VALIDMAX "valid_max"
#define VALIDRANGE "valid_range"

int main(void)
{
    herr_t  status= FAIL;
    hid_t   swid, file, dataset, att;  /* file, dataset, attribute handles */
    hid_t   dtype, dspace;
    hsize_t size[2]= {NX1, NY1}, temp_size;
    char  *degrees_east= "degrees_east", *degrees_north= "degrees_north",
            *kelvin= "kelvin", *latitude= "latitude",
            *longitude= "longitude", *temperature= "temperature",
             *coorlist[2]= {"lat", "lon"};
    float    temp_array[NX1][NY1];
    float  lat_array[NX1][NY1];
    float   lon_array[NX1][NY1];
    float value[1];
    int     i, j;

/* PART 1: temp */

   /* values between a[60][*] and a[120][*] is around 300.0 */
    for(i=60; i<=120; i++)
        for(j=0; j<NY1; j++)
            temp_array[i][j]= 300.0;

   /* values between a[0][*] and a[59][*], a[121][*] and a[179][*]
        is around 280.0 */
    for(i=0; i<60; i++)
        for(j=0; j<NY1; j++)
            temp_array[i][j]= 280.0;

    for(i=121; i<NX1; i++)
        for(j=0; j<NY1; j++)
            temp_array[i][j]= 280.0;

    /* a few random values should be -999.0; ~5% */
/*
    for(i=0; i<NX1*.05; i++)
        for(j=0; j<NY1*.05; j++)
            temp_array[i][j]= rand() %
*/
    for(i=0; i<NX1; i++)
        lat_array[i][0]= -90.0;
    for(i=0; i<NX1; i++)
        for(j=1; j<NY1; j++)
            lat_array[i][j]= lat_array[i][j-1]+1.0;

    for(j=0; j<NY1; j++)
        lon_array[0][j]= -180.0;
    for(j=1; j<NY1; j++)
        for(i=1; i<NX1; i++)
            lon_array[i][j]= lon_array[i-1][j]+1.0;

    /* Open a new HDF-EOS swath file, "Swath.he5" */ 
    file = HE5_SWopen("swath.he5", H5F_ACC_TRUNC); 

    /* Create the swath, "Swath1", within the file */ 
    swid = HE5_SWcreate(file, "Swath1"); 

    /* Define dimensions and specify their sizes */ 
    status = HE5_SWdefdim(swid, "GeoXtrack", NX1); 
    status = HE5_SWdefdim(swid, "GeoTrack", NY1); 

    /* fill value MUST be called before defining the data field */
    value[0]= -999.0;
    HE5_SWsetfillvalue(swid, TEMP, H5T_NATIVE_FLOAT, &value[0]);

    /* Define data fields (datasets) */
    HE5_SWdefdatafield(swid, TEMP, "GeoTrack,GeoXtrack", NULL, 
            H5T_NATIVE_FLOAT, 0);
 
    status = HE5_SWwritefield(swid, TEMP, NULL, NULL, NULL, temp_array);

/*************   add cf attributes **********/

/* part 1: temperature */
    /* add units */
    size[0]= strlen(kelvin);
    status= HE5_SWwritelocattr(swid, TEMP, UNITS, H5T_C_S1, 
            &size[0], (void*)kelvin);

    /* add long_name */
    size[0]= strlen(temperature);
    status= HE5_SWwritelocattr(swid, TEMP, LONGNAME, H5T_C_S1, 
            &size[0], (void*)temperature);

    /* add coordinates attribute */
    size[0]= 2;
    dtype= H5Tcopy(H5T_C_S1);
    status= H5Tset_size(dtype, H5T_VARIABLE);
    status= HE5_SWwritelocattr(swid, TEMP, COORDINATES, dtype,
            &size[0], coorlist);

    /* add valid_range */
    size[0]= 2;
    value[0]= 0;
    value[1]= 400;
    status= HE5_SWwritelocattr(swid, TEMP, VALIDRANGE, H5T_NATIVE_FLOAT, 
            &size[0], value);
    
    /* add valid_min */
    size[0]= 1;
    status= HE5_SWwritelocattr(swid, TEMP, VALIDMIN, H5T_NATIVE_FLOAT, 
            &size[0], &value[0]);

    /* add valid_min */
    status= HE5_SWwritelocattr(swid, TEMP, VALIDMAX, H5T_NATIVE_FLOAT, 
            &size[0], &value[1]);

    /* add add_offset */
    value[0]= 10;
    status= HE5_SWwritelocattr(swid, TEMP, ADDOFFSET, H5T_NATIVE_FLOAT, 
            &size[0], &value[0]);

    /* add scale_factor */
    value[0]= 1.01;
    status= HE5_SWwritelocattr(swid, TEMP, SCALEFACTOR, H5T_NATIVE_FLOAT, 
            &size[0], &value[0]);




    /* define dimensions */
    status = HE5_SWdefgeofield(swid, "lon", "GeoTrack,GeoXtrack", 
            NULL, H5T_NATIVE_FLOAT, 0);
    status = HE5_SWdefgeofield(swid, "lat", "GeoTrack,GeoXtrack", 
            NULL, H5T_NATIVE_FLOAT, 0);

/* part 2: latitude */

    /* write latitude dataset */
    status= HE5_SWwritefield(swid, LAT, NULL, NULL, 
            NULL, lat_array);

    /* add units */
    size[0]= strlen(degrees_north);
    status= HE5_SWwritelocattr(swid, LAT, UNITS, H5T_C_S1,
            &size[0], (void*)degrees_north);

    /* add long_name */
    size[0]= strlen(latitude);
    status= HE5_SWwritelocattr(swid, LAT, LONGNAME, H5T_C_S1,
            &size[0], (void*)latitude);

    /* add standard_name */
    status= HE5_SWwritelocattr(swid, LAT, STDNAME, H5T_C_S1,
            &size[0], (void*)latitude);

/* part 3: longitude */

    status= HE5_SWwritefield(swid, LON, NULL, NULL, 
            NULL, lon_array);

    /* add units */
    size[0]= strlen(degrees_east);
    status= HE5_SWwritelocattr(swid, LON, UNITS, H5T_C_S1,
            &size[0], (void*)degrees_east);

    /* add long_name */
    size[0]= strlen(longitude);
    status= HE5_SWwritelocattr(swid, LON, LONGNAME, H5T_C_S1,
            &size[0], (void*)longitude);

    /* add standard_name */
    status= HE5_SWwritelocattr(swid, LON, STDNAME, H5T_C_S1,
            &size[0], (void*)longitude);


    /* Close the swath interface */ 
    status = HE5_SWdetach(swid); 
    /* Close the swath file */ 
    status = HE5_SWclose(file); 

return 0;
}

File: eos5/swath.f
c created by:  Isayah Reed <ireed@hdfgroup.org>
c              May  4, 2011
c  
c  This is a simple program that demonstrates how to use HDF-EOS5 to add CF
c  attributes to an h5 file. Creates a swath called Swath1 and adds the
c  temperature dataset with 2 dimensions: latitude and longitude. The
c  temperature dataset has the CF attributes: units, long_name, _FillValue,
c  coordinates, valid_min, valid_max, valid_range, scale_factor, add_offset.
c  Latitude and longitude have the CF attributes units and long_name. Outputs
c  data to general.h5
c  
c Compile:
c  $(HDF5_DIR)/bin/h5fc swath.f -I$(EOS_DIR)/include -L$(EOS_DIR)/lib -lhe5_hdfeos -lGctp

      program     swath

      implicit    none
      include     'hdfeos5.inc'

      integer     status
      integer     he5_swopen
      integer     he5_swcreate
      integer     he5_swsetfill
      integer     he5_swdefgfld
      integer     he5_swdefdfld
      integer     he5_swdefdim
      integer     he5_swwrfld
      integer     he5_swwrlattr
      integer     he5_swdetach
      integer     he5_swclose
      integer     swfid, swid

      integer*4   dtrack, strlen(2), value, NX1, NY1
      integer*4   start(2), stride(2), edge(2)
      parameter   (NX1=180, NY1=360)
      real   lon_array(NX1,NY1), lat_array(NX1,NY1)
      real   temp_data(NY1,NX1)
      character*50 attr(2)
      integer     i, j

      integer     FAIL
      parameter   (FAIL=-1)

c     Open the HDF-EOS file, "swath.he5" using "READ/WRITE" access code
c     -----------------------------------------------------------------
      swfid = he5_swopen('swath.he5',HE5F_ACC_TRUNC)

      swid = he5_swcreate(swfid, "Swath1")

c   Initialize temperature data.
      do i = 1, 60
          do j = 1, NY1
              temp_data(i,j) = 280.0
          enddo
      enddo
      do i = 61, 120
          do j = 1, NY1
              temp_data(i,j) = 300.0
          enddo
      enddo
      do i = 121, NX1
          do j = 1, NY1
              temp_data(i,j) = 280.0
          enddo
      enddo

c   Initialize latitude data.
        do i=1, NX1
            lat_array(i,1)= -90.0;
        enddo
        do i=1, NX1
            do j=2, NY1
                lat_array(i,j)= lat_array(i,j-1)+1.0;
            enddo
        enddo

c   Initialize longitude data.
        do j=1, NY1
            lon_array(1,j)= -180.0;
        enddo
        do j=2, NY1
            do i=2, NX1
                lon_array(i,j)= lon_array(i-1,j)+1.0;
            enddo
        enddo

c   define latitude and longitude dimensions
      dtrack = NX1
      status = he5_swdefdim(swid, "GeoXtrack", dtrack)

      dtrack = NY1
      status = he5_swdefdim(swid, "GeoTrack", dtrack)


c     Define Geolocation and Data fields
c     ----------------------------------
c     ---------------------------------------------------------------
c     We define six fields.  The first three, "Time", "Longitude"
c	  and "Latitude" are geolocation fields and thus we use the
c	  geolocation dimensions "GeoTrack" and "GeoXtrack" in the field
c	  definitions.  We also must specify the data type using the
c	  standard HDF data type codes.  In this example the geolocation
c	  are 4-byte (32 bit) floating point numbers.
c     
c	  The next three fields are data fields.  Note that either
c	  geolocation or data dimensions can be used. 
c     ---------------------------------------------------------------

c   set _FillValue
        value= -999
        status = he5_swsetfill(swid, "temp", HE5T_NATIVE_FLOAT, value)            

c   define data fields
        status = he5_swdefdfld(swid, "temp", "GeoTrack,GeoXtrack",
     1      " ", HE5T_NATIVE_FLOAT, 0)     

c   write the temperature dataset
        start(1)= 0
        start(2)= 0
        stride(1)= 1
        stride(2)= 1
        edge(1)= NY1
        edge(2)= NX1
        status= he5_swwrfld(swid, "temp", start, stride, edge, 
     1          temp_data)

c   add long_name attribute
        status = he5_swwrlattr(swid,"temp","long_name",
     1      HE5T_NATIVE_CHAR,11,"temperature")

c   add units attribute
        strlen(1)= 6
        status = he5_swwrlattr(swid,"temp","units",
     1      HE5T_NATIVE_CHAR,strlen,"kelvin")

c   add valid_min attribute
        strlen(1)= 1
        status = he5_swwrlattr(swid,"temp","valid_min",
     1      HE5T_NATIVE_FLOAT,strlen,0.0)

c   add valid_max attribute
        status = he5_swwrlattr(swid,"temp","valid_max",
     1      HE5T_NATIVE_FLOAT,strlen,400.0)

c   add valid_range attribute
        strlen(1)= 2
        edge(1)= 275
        edge(2)= 305
        status = he5_swwrlattr(swid,"temp","valid_range",
     1      HE5T_NATIVE_FLOAT,strlen,edge)

c   add add_offset attribute
        status = he5_swwrlattr(swid,"temp","add_offset",
     1      HE5T_NATIVE_FLOAT,1,1)

c   add scale_factor
        status = he5_swwrlattr(swid,"temp","scale_factor",
     1      HE5T_NATIVE_FLOAT,1,1.01)

c    add coordinates attribute
        strlen(1)= 3
        strlen(2)= 3
        attr(1)= "lat"
        attr(2)= "lon"
        status = he5_swwrlattr(swid,"temp","coordinates",
     1      HE5T_NATIVE_CHAR,strlen,attr)



c   define latitude dimension
        status = he5_swdefgfld(swid, "lat", "GeoTrack,GeoXtrack",
     1      " ", HE5T_NATIVE_FLOAT, 0)
c   define longitude dimension
        status = he5_swdefgfld(swid, "lon", "GeoTrack,GeoXtrack",
     1      " ", HE5T_NATIVE_FLOAT, 0)




c   write latitude dataset
        status= he5_swwrfld(swid, "lat", start, stride, edge, 
     1      lat_array)

c   add long_name attribute
        strlen(1)= 8
        attr(1)= "latitude"
        status = he5_swwrlattr(swid,"lat","long_name",
     1      HE5T_NATIVE_CHAR,strlen,attr)

c   add units attribute
        status = he5_swwrlattr(swid,"lat","units",
     1      HE5T_NATIVE_CHAR,13,"degrees_north")

        status= he5_swwrfld(swid, "lon", start, stride, edge, 
     1      lon_array)
            




c   write longitude dataset
        edge(1)= NX1
        edge(2)= NY1
        status= he5_swwrfld(swid, "lon", start, stride, edge,
     1      lon_array)

c   add long_name attribute
        strlen(1)= 9
        attr(1)= "longitude"
        status = he5_swwrlattr(swid,"lon","long_name",
     1      HE5T_NATIVE_CHAR,strlen,attr)

c   add units attribute
        status = he5_swwrlattr(swid,"lon","units",
     1      HE5T_NATIVE_CHAR,12,"degrees_east")


      
c     Detach from the swath
c     ---------------------      
      status = he5_swdetach(swid)
      
c     Close the file
c     --------------      
      status = he5_swclose(swfid)

      stop
      end




